volatility3/framework/plugins/linux/vmayarascan.py

```diff
@@ -66,29 +71,36 @@ def _generator(self):
             # get the proc_layer object from the context
             proc_layer = self.context.layers[proc_layer_name]
 
-            for start, end in self.get_vma_maps(task):
-                for match in rules.match(
-                    data=proc_layer.read(start, end - start, True)
-                ):
-                    if yarascan.YaraScan.yara_returns_instances():
-                        for match_string in match.strings:
-                            for instance in match_string.instances:
-                                yield 0, (
-                                    format_hints.Hex(instance.offset + start),
-                                    task.UniqueProcessId,
-                                    match.rule,
-                                    match_string.identifier,
-                                    instance.matched_data,
-                                )
-                    else:
-                        for offset, name, value in match.strings:
-                            yield 0, (
-                                format_hints.Hex(offset + start),
-                                task.tgid,
-                                match.rule,
-                                name,
-                                value,
-                            )
+            vma_maps = list(self.get_vma_maps(task))
+            insane_vma_maps = [
+                start for (start, size) in vma_maps if size > sanity_check
+            ]
+            for start in insane_vma_maps:
+                vollog.debug(f"VMA at 0x{start:x} over sanity-check size, not scanning")
+
+            if not vma_maps:
+                vollog.warning(f"No VMAs were found for task {task.pid}, aborting")
+                continue
+
+            max_vma_size: int = max(
+                [size for (start, size) in vma_maps if size <= sanity_check]
+            )
```

Comment:
There are a couple of issues with this implementation:

- Whole List Used for Scanning: The scan operation ends up using the entire list. Please refer to the comment on the proc_layer.scan() call below for more details.

- Incorrect Max VMA Size Calculation: The `max_vma_size` calculation includes oversized VMAs that should be ignored.

Also, there are 4 "for" statements or loops over the same group of data and 3 list being instantiated in this small piece of code. I would recommend refactoring this to loop over the data just once. For example, you could try something like this (make sure to test it):
```python
max_vma_size: int = 0
vma_maps = []
for vma_map in self.get_vma_maps(task):
    start, size = vma_map
    if size > sanity_check:
        vollog.debug(f"VMA at 0x{start:x} over sanity-check size, not scanning")
        continue 
    elif size > max_vma_size:
        max_vma_size = size

    vma_maps.append(vma_map)

if not vma_maps:
    vollog.warning(f"No VMAs were found for task {task.pid}, aborting")
    continue
```

Reply:
I think I was mid-refactoring it to use either the `insane_vma_size` or `max_vma_size`, but I think I got distracted.  Thanks for catching it in review!  I'll get some changes in place...